_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+This means interacting asynchronously and outside the user interaction where this is applicable.
+
+Going this route you will soon ask yourself how to test such asynchronous interactions.
+[Spring Cloud Contract](https://cloud.spring.io/spring-cloud-contract/) offers great support for such scenarios.
+
+Let's look at the problem using a simple application consisting of two services and see how Spring Cloud Contract can help us to improve testing the asynchronous interaction between them.
+
+## The Sample Application
+
+Our application consists of two services:
+
+- the product service is responsible for managing product data
+- the checkout service offers cart-related capabilities
+
+The checkout service cannot fulfill the task of adding a certain quantity of a product to a cart without product data.
+So the checkout service listens to product events to hold a copy of relevant data of each product.
+
+This approach allows us to satisfy all user requests in the checkout service from its local database which should be a lot faster than having to ask the product service for product data while the user is waiting for a reply.
+We also reduce runtime dependencies between our services.
+
+For the sample application, we use [Spring Boot](http://projects.spring.io/spring-boot/),  [RabbitMQ](https://www.rabbitmq.com/) as the message broker, and [Spring Amqp](http://projects.spring.io/spring-amqp/) to interact with it from our application.
+
+The code referenced in this post is available on [github](https://github.com/mduesterhoeft/testing-asynchonous-interactions). It is advisable to make yourself familiar with this code before reading on.
+
+Whenever we create a [product](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/product-service/src/main/java/com/epages/product/Product.java) an event is [published](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/product-service/src/main/java/com/epages/product/ProductEventHandler.java) which the checkout service [picks up](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/java/com/epages/checkout/ProductSubscriber.java).
+It will then store [its view on the product](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/java/com/epages/checkout/ProductRef.java) in the checkout database.
+
+## A primitive test is better than nothing
+
+How should we test this interaction between the product and the checkout service?
+Of course, an integration test would be possible.
+But we favor fast feedback and ideally want issues to become apparent as early as possible and without having to start all our services.
+So we strive for a solution that allows us to test the interactions on the checkout service in isolation.
+We do not want to run the product service and the message broker to be able to test it.
+We call such a test a local integration test.
```

Comment:
This is what we call a local integration test. (avoids another "we" at the beginning)